src/basic/stat-util.c

```diff
@@ -213,15 +213,47 @@ int fd_is_network_fs(int fd) {
 }
 
 int fd_is_network_ns(int fd) {
+        struct statfs s;
         int r;
 
-        r = fd_is_fs_type(fd, NSFS_MAGIC);
-        if (r <= 0)
-                return r;
+        /* Checks whether the specified file descriptor refers to a network namespace. On old kernels there's no nice
+         * way to detect that, hence on those we'll return a recognizable error (EUCLEAN), so that callers can handle
+         * this somewhat nicely.
+         *
+         * This function returns > 0 if the fd definitely refers to a network namespace, 0 if it definitely does not
+         * refer to a network namespace, -EUCLEAN if we can't determine, and other negative error codes on error. */
+
+        if (fstatfs(fd, &s) < 0)
+                return -errno;
+
+        if (!is_fs_type(&s, NSFS_MAGIC)) {
+                /* On really old kernels, there was no "nsfs", and network namespace sockets belonged to procfs
+                 * instead. Handle that in a somewhat smart way. */
+
+                if (is_fs_type(&s, PROC_SUPER_MAGIC)) {
+                        struct statfs t;
+
+                        /* OK, so it is procfs. Let's see if our own network namespace is procfs, too. If so, then the
+                         * passed fd might refer to a network namespace, but we can't know for sure. In that case,
+                         * return a recognizable error. */
+
+                        if (statfs("/proc/self/ns/net", &t) < 0)
+                                return -errno;
+
+                        if (s.f_type == t.f_type)
+                                return -EUCLEAN; /* It's possible, we simply don't know */
```

Comment:
I'm not sure I understand why `is_fs_type(&s, PROC_SUPER_MAGIC) return -EUCLEAN` isn't enough. Do you have some kind of a failure mode in mind that I'm missing?

Reply:
More generally, why is `fd_is_network_ns` necessary at all? It seems to be possible to just improve the function used for joining namespaces so that it would log something meaningful when `setns` returns `-EINVAL`.

Reply:
That's not to say that I'm strongly opposed to `fd_is_network_ns`. It's just that it's the second time `nspawn` has broken due to it.

Reply:
> I'm not sure I understand why is_fs_type(&s, PROC_SUPER_MAGIC) return -EUCLEAN isn't enough. Do you have some kind of a failure mode in mind that I'm missing?

The ambiguity exists only on old kernels. On new kernels we can reliably discern namespacing fds from non-namespacing fds. i.e. on new kernels we do know that a procfs fd cannot be a namespacing fd, while on old kernels that's possible. The additional check hence verifies if our own netnsfd is also procfs. if it is then we are on an old kernel. if not, then we are on a new kernel and can make a clear decision.

Reply:
> More generally, why is `fd_is_network_ns` necessary at all? It seems to be possible to just improve the function used for joining namespaces so that it would log something meaningful when `setns` returns `-EINVAL`.

Strictly necessary, no. But we invoke the setns() really far down the setup process. And given that the fd passed in is "input" and that we better validate input at the incoming boundary, early on I think it's wise to do so if we easily can.

Reply:
> That's not to say that I'm strongly opposed to `fd_is_network_ns`. It's just that it's the second time `nspawn` has broken due to it.

hmm? what's the first one?

Reply:
> hmm? what's the first one?

The first one was https://github.com/systemd/systemd/issues/8543.

> And given that the fd passed in is "input" and that we better validate input at the incoming boundary, early on I think it's wise to do so if we easily can.

That's true, but I wouldn't say we can do it easily and reliably. I wouldn't say that it isn't going to stop working when something changes in the kernel either :-) Then again, I'm not strongly against it.

Reply:
> The first one was #8543.

Well, that was just me being stupid. There's nothing unreliable in the kernel in that regard, that was solely and alone my own fault for writing bad code.

This new issue is different though: it's a compat issue with old kernels. Given that features once added to the kernel remian quite stable we shouldn't have a problem in the future I presume